libethereum/ExtVM.cpp

```diff
@@ -95,30 +95,78 @@ void go(unsigned _depth, Executive& _e, OnOpFunc const& _onOp)
 
 bool ExtVM::call(CallParameters& _p)
 {
-	Executive e(m_s, envInfo(), m_sealEngine, depth + 1);
+	m_successfulCalls.emplace_back(m_s, envInfo(), m_sealEngine, depth + 1);
+	auto& e = m_successfulCalls.back();
 	if (!e.call(_p, gasPrice, origin))
 	{
 		go(depth, e, _p.onOp);
 		e.accrueSubState(sub);
 	}
 	_p.gas = e.gas();
 
-	return !e.excepted();
+	if (e.excepted())
+	{
+		m_successfulCalls.pop_back();
+		return false;
+	}
+	return true;
 }
 
 size_t ExtVM::codeSizeAt(dev::Address _a)
 {
 	return m_s.codeSize(_a);
 }
 
+void ExtVM::setStore(u256 _n, u256 _v)
+{
+	if (!m_origStorage.count(_n))
+	{
+		m_origStorage.emplace(_n, store(_n));
+		clog(ExecutiveWarnChannel) << "ORIG STORAGE " << myAddress << _n << _v;
+	}
+	m_s.setStorage(myAddress, _n, _v);
+}
+
 h160 ExtVM::create(u256 _endowment, u256& io_gas, bytesConstRef _code, OnOpFunc const& _onOp)
 {
-	Executive e(m_s, envInfo(), m_sealEngine, depth + 1);
+	// Every CREATE increases this account nonce, no matter if it succeeds.
+	++m_nonceInc;
+
+	m_successfulCalls.emplace_back(m_s, envInfo(), m_sealEngine, depth + 1);
+	auto& e = m_successfulCalls.back();
 	if (!e.create(myAddress, _endowment, gasPrice, io_gas, _code, origin))
 	{
 		go(depth, e, _onOp);
 		e.accrueSubState(sub);
 	}
 	io_gas = e.gas();
+	if (!e.newAddress())
+		m_successfulCalls.pop_back();
 	return e.newAddress();
 }
+
+void ExtVM::revert()
+{
+	clog(ExecutiveWarnChannel) << "Reverting " << myAddress;
+	for (auto it = m_successfulCalls.rbegin(); it != m_successfulCalls.rend(); ++it)
```

Comment:
I'm not sure if it is a good idea keeping a full executive around. Perhaps it is just better to store a vector of operations to revert and append the revert-vector of each successful call to the parent vector. Here, "operation to revert" is just either "(store, 0x123, 0x45)", "incnonce", "changeBalance, 0x123, 0x34", and whatever else we have.

Another solution would be to create a separate class "ReversionLog" that contains all the information needed in here and might refer to sub-logs.